File: spring-multisport/src/main/java/com/demo/multisport/entities/center/Center.java
package com.demo.multisport.entities.center;

import com.demo.multisport.entities.Plan;
import com.demo.multisport.entities.page.Comment;
import com.demo.multisport.entities.page.Rating;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.util.Set;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "center_type", discriminatorType = DiscriminatorType.STRING)
@Table(name = "centers")
@Getter
@Setter
public abstract class Center implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @NotNull
    @Size(min = 4, max = 20)
    @Column(nullable = false, length = 20)
    private String name;

    @NotNull
    @Size(min = 10, max = 40)
    @Column(nullable = false, length = 40, unique = true)
    private String address;

    @NotNull
    @Size(min = 4)
    @Column(columnDefinition = "TEXT NOT NULL")
    private String description;

    // check length
    @NotNull
    @ElementCollection
    @Column(columnDefinition = "VARCHAR(512) NOT NULL")
    private Set<String> pictures;

    @JsonBackReference
    @ManyToMany(mappedBy = "centers")
    private Set<Plan> plans;

    @JsonManagedReference
    @OneToMany(mappedBy = "center")
    private Set<Comment> comments;

    @OneToOne(cascade = CascadeType.ALL, optional = false)
    @JoinColumn(name = "rating_id", referencedColumnName = "id",
            foreignKey = @ForeignKey(name = "FK_CENTER_RATING"))
    @JsonManagedReference
    private Rating rating;

    protected Center() { }

}

File: angular-multisport/src/app/models/centers/Center.ts
import { CenterType } from "./typecenter/CenterType"
import { Comment } from "../page/Comment"
export abstract class Center {
    private readonly _centerType: CenterType
    private _id: number
    private _name: string
    private _address: string
    private _description: string
    private _pictures: string[] = []
    private _rating: number
    private _comments: Comment[]

    constructor (id: number, name: string, address: string, description: string, pictures: string[], centerType: CenterType, rating: number, comments: Comment[] ) {
        this._id = id
        this._name = name
        this._address = address
        this._description = description
        this._pictures = pictures
        this._centerType = centerType
        this._rating = rating
        this._comments = comments
    }

    public get rating(): number {
        return this._rating
    }

    public get id(): number {
        return this._id
    }

    public get centerType(): CenterType {
        return this.centerType
    }

    public get pictures(): string[] {
        return this._pictures
    }

    public get name(): string {
        return this._name
    }

    public get address(): string {
        return this._address
    }

    public get description(): string {
        return this._description
    }
    
    public get comments(): Comment[] {
        return this._comments
    }
}